src/ITensorChainRules/ITensorChainRules.jl

```diff
@@ -326,6 +319,52 @@ function ChainRulesCore.rrule(::typeof(apply), x1::Vector{ITensor}, x2::MPS; kwa
   return y, apply_pullback
 end
 
+function ChainRulesCore.rrule(::typeof(apply), x1::Vector{ITensor}, x2::MPO; kwargs...)
+  apply_dag = kwargs[:apply_dag]
+  N = length(x1) + 1
+  # Apply circuit and store intermediates in the forward direction
+  x1x2 = Vector{MPO}(undef, N)
+  x1x2[1] = x2
+  for n in 2:N
+    x1x2[n] = apply(x1[n - 1], x1x2[n - 1]; move_sites_back=true, kwargs...)
+  end
+  y = x1x2[end]
+  function apply_pullback(ȳ)
+    x1x2dag = dag.(x1x2)
+    x1dag_ȳ = Vector{MPO}(undef, N)
+    x1dag_ȳ[end] = ȳ
+    x1dag = [swapprime(dag(x), 0 => 1) for x in x1]
+    for n in (N - 1):-1:1
+      x1dag_ȳ[n] = apply(x1dag[n], x1dag_ȳ[n + 1]; move_sites_back=true, kwargs...)
+    end
+
+    x̄1 = similar(x1)
+    for n in 1:length(x1)
+      if iseven(length(inds(x1[n])))
```

Comment:
Could you explain what this is doing?

Reply:
This is just checking whether you have a noisy gate, which are currently not differentiable. I left a comment on top of this `if` statement but got deleted somehow between commits

Reply:
I see, thanks.

Reply:
Wouldn't that be an issue with applying a noisy gate to an MPS in general, and nothing particular to defining the `rrule`?

Reply:
To be clear, applying a noisy gate to an MPS right now is not well defined behavior.

Reply:
Yes correct